src/block/reader.rs

```diff
@@ -382,6 +382,55 @@ impl<R: ChunksReader> SequentialBlockDecompressor<R> {
     }
 }
 
+// note: this is just pseudocode, it will be split up into a lot of separate pieces to allow for customization
+fn read_all_blocks_fully_parallel<Read, PixelBlock>(
+    mut create_reader_at_byte: impl Sync + FnMut(u64) -> Result<Read>,
+    pedantic: bool
+)
+    -> Result<impl Iterator<Item = Result<PixelBlock>>>
+    where Read: Read, PixelBlock: PixelsFromDecompressedBlock
+{
+    // TODO do not create a buffer for in-memory vectors
+    let mut read_buffered = Tracking::new(BufRead::new(create_reader_at_byte(0)?));
+    let meta = MetaData::read_from_buffered(&mut read_buffered, pedantic)?;
+
+    let offset_tables = MetaData::read_offset_tables(&mut read_buffered, &meta.headers)?;
+
+    if pedantic {
+        validate_offset_tables(
+            meta.headers.as_slice(), &offset_tables,
+            read_buffered.byte_position()
+        )?;
+    }
+
+    std::mem::drop(read_buffered); // TODO re-use when single threaded?
+
+    let meta = Arc::new(meta);
+    let thread_pool = threadpool::Builder::new().build();
+    let (sender, receiver) = ::flume::unbounded(); // TODO bounded? bounded to threadpool size?
+
+    for &chunk_index in offset_tables.iter().flatten() {
+        let meta = meta.clone();
+
+        thread_pool.spawn(move ||{
+            let meta = meta.as_ref();
+
+            let result = {
+                // TODO might want to do this on main thread such that the closure does not have to be sync?
+                let file_handle = create_reader_at_byte(chunk_index)?;
```

Comment:
I'm pretty sure you cannot read from the same file descriptor from multiple threads at once, and repeated calls to `open()` might be expensive.

What I'd do is have each thread have its own `file_handle`, all pointing to the same underlying file.

I wonder what happens if you use `thread_local` - will it complain that the data from it cannot be used on other threads, or will it create the `thread_local` on all threads and let you initialize it from the closure?

Reply:
`create_reader` was intended to call `open` each time, inside each thread. this way, each thread has its own `file_handle`. alternatively, the closure can also open the file once and then call `try_clone` on the file handle. the closure has the responsibility that each call opens the same file at the specified location, but everything else can be changed. 

using `thread_local` would be something the closure could also do, I assume.

Reply:
update: `try_clone` will not have it's own seek position :(

[docs](https://doc.rust-lang.org/std/fs/struct.File.html#method.try_clone)

Reply:
for the initial implementation, I would prefer to open multiple file handles, and then benchmark

Reply:
Fair enough :+1:

Reply:
we could also try using `try_clone`, which has no separate seek position, but then use `file.read_exact_at(offset)` which does not modify the seek position in the file handle. could it be faster?

see https://docs.rs/read_write_at/0.1.0/read_write_at/trait.ReadAt.html#method.read_exact_at

Reply:
Perhaps! I've never used that API so you'll just have to try it.